qcodes/tests/test_autoloadable_channels.py

```diff
@@ -0,0 +1,186 @@
+            if ans is not None:
+                break
+
+        if ans is None:
+            raise ValueError(f"Command {cmd} unknown")
+
+        args = ans.groups()
+        return self._command_dict[key](*args)
+
+
+class TestInstrumentBackend(DummyBackendBase):
+    def __init__(self):
+        super().__init__()
+        self._channel_catalog = ["1", "2", "4"]
+
+        self._command_dict = {
+            r":INST:CHN(\d):HLO": lambda chn: f"Hello from channel {chn}",
+            r":INST:CHN:ADD (\d)": self._channel_catalog.append,
+            r":INST:CHN:DEL (\d)": self._channel_catalog.remove,
+            r":INST:CHN:CAT": lambda: ",".join([str(i) for i in self._channel_catalog])
+        }
+
+
+class SimpleTestChannel(AutoLoadableInstrumentChannel):
+    """
+    A channel to test if we can create and delete channel instances
+    """
+
+    @classmethod
+    def _discover_from_instrument(
+            cls, parent: Instrument, **kwargs) -> List[dict]:
+        """
+        New channels need `name` and `channel` keyword arguments.
+        """
+        channels_str = parent.channel_catalog()
+        kwarg_list = [
+            {"name": f"channel{i}", "channel": int(i)}
+            for i in channels_str.split(",")
+        ]
+
+        return kwarg_list
+
+    @classmethod
+    def get_new_instance_kwargs(cls, parent: Instrument, **kwargs) -> dict:
+        """
+        Find the smallest channel number not yet occupied
+        """
+        channels_str = parent.channel_catalog()
+        existing_channels = [int(i) for i in channels_str.split(",")]
+
+        new_channel = 1
+        while new_channel in existing_channels:
+            new_channel += 1
+
+        kwargs = {
+            "name": f"channel{new_channel}",
+            "channel": new_channel
+        }
+
+        return kwargs
+
+    def __init__(
+            self,
+            parent: Union[Instrument, 'InstrumentChannel'],
+            name: str,
+            channel: int,
+            existence: bool = False,
+            channel_list: 'AutoLoadableChannelList' = None,
+    ) -> None:
+
+        super().__init__(parent, name, existence, channel_list)
+        self._channel = channel
+
+        self.add_parameter(
+            "hello",
+            get_cmd=f":INST:CHN{self._channel}:HLO"
+        )
+
+    def _create(self)->None:
+        """Create the channel on the instrument"""
+        self.parent.write(f":INST:CHN:ADD {self._channel}")
+
+    def _delete(self)->None:
+        """Remove the channel from the instrument"""
+        self.write(f":INST:CHN:DEL {self._channel}")
+
+
+class TestInstrument(Instrument):
+    """
+    This dummy instrument allows the creation and deletion of
+    channels
+    """
+
+    def __init__(self, name):
+        super().__init__(name)
+
+        self._backend = TestInstrumentBackend()
+
+        self.add_parameter(
+            "channel_catalog",
+            get_cmd=":INST:CHN:CAT",
+        )
```

Comment:
to my understanding, this parameter belongs to `DummyInstrumentWithChannelList`, right?

Reply:
to my understanding it does not, because it is not used in this class. This parameter is only used by the channels that are created in `DummyInstrumentWithChannelList`. `TestInstrument` as an entity can exist perfectly without the notion of the channels.

I made this comment because I was a bit confused when reading the code why you create `DummyInstrumentWithChannelList` __and__ `TestInstrument`. Then I understood that `TestInstrument` was made to hide the "mock" backend. But then having that parameter in `TestInstrument` does not make much sense, but makes a lot of sense in `DummyInstrumentWithChannelList` since channels rely on it.